### DISCUSSION TITLE
Avoiding pump stop when njspc container is stopped?

### DISCUSSION THREAD

**Discussion by MaxVonEvil** [tagyoureit/nodejs-poolController#546] [Q&A]
Hi @tagyoureit et. al
TLDR; poolpump receives a command to stop when njsPC docker container is stopped. Can that be avoided through configuration?

Since upgrading to 7.7.0 I've noticed something different in my setup: Once a week my pool guy comes around to clean the pool, he hooks up his own gear and puts the pump on speed4 set for 3100rpm. He does this manually as he needs to stop the pump to stick the vac hose in the skimmer's intake. After he's left (trust me, my hpuse knows ;) I take back control of the pump with NjsPC.

As you know when NjsPC takes over the pump, the local panel on a Pentair Intelliflo VS reads; "DISPLAY NOT ACTIVE" and the keys are disabled, i.e no manual control for Mr. Poolguy

What I used to do was once a week stop njspc from running in the morning before he arrives so after 30 sec or so, the local programming on the pump would kick in.

However somewhere between v7.7.0 and 5.x it seems that NjsPC's behavior changes. If I do a 'docker stop poolcontroller' it seems a stop command or rpm=0 (same difference) is sent to the pump, rather than let it resume it's on-board programmed schedule

Here comes my conundrum: Is it possible to avoid that behavior? I noticed that if I pull the data plug physically from the pump, it does what it used to do; fallback to local scheduling. So if all else fails I suppose I could let my RPi control a relaymodule on a GPIO port which cuts the data connection physically. But it seems like a hack even McGyver would facepalm over, so before I break out the soldering iron, I figured I'd ask the experts that the best approach to getting the behavior back

PS: Please don't suggest that poolguy has to pull the plug, press a IoT button etc, he's an old guy set in his ways, yet does a bang-up good job overallso I'm just trying to make sure the automation gets out of his way when he's here.

Hope there is an easy solution to this.
Thanks,
Max

**Comment by rstrouse** [comment]
If you have the latest njsPC and dashPanel you can click on the ready icon and set the panel to timeout or device mode.  This will release the pump.

**Reply by rstrouse** [reply]
See this thread.  https://github.com/tagyoureit/nodejs-poolController/discussions/508

**Reply by tagyoureit** [reply]
The only change that was made recently was to affect the pumps was to PM2 shutdown.  We noticed PM2 was killing the njsPC before the pumps were properly shut down.  I don't think that would affect docker and can't think of any other changes that would have affected the pumps in the recent past.

**Reply by MaxVonEvil** [reply]
I see, thanks gents. 

@tagyoureit - PM2 killing it might have been what I was inadvertently relying on before and docker probably isn't doing that. It would explain why my pump now stops when the docker container is shut down.

@rstrouse, reading the thread, I understand the feature on the statusbar you mentioned requires relay equipment manager installed. Is there a dockerized version of REM on the drawingboard? After having had multiple node version conflicts I much prefer having anything nodejs encapsulated in containers to keep things tidy.

That said, as I already have all my other equipment, incl. Pi running njspc under firm control of HA/NodeRed, it seems a bit much to install REM on top to get that single feature. Is there any chance that you guys might consider a tiny flag in the poolcontroller's config.json, something like "DontStopPumpAtShutDown" : true 

-Pretty please? :)

**Reply by rstrouse** [reply]
It does not require REM.  The user simply added hardware push buttons and LEDs at the pad that would toggle the system at the pad.  They used REM to manage the hardware LEDs and buttons.  

If you are running the current repo click on the status bar and the feature will be there.

**Comment by MaxVonEvil** [comment]
Hmm, curious and curiouser. 
I am running 7.7.0 of both njspc and dashpanel as you can see
![image](https://user-images.githubusercontent.com/19863636/179617850-8bae8faa-94be-4db2-9f07-ae6dfca7f342.png)
However, when I click anywhere between the burger menu on the left and the gears on the right nothing happens? I've checked/disabled adblockers etc, but no cigar? I wonder what I might be missing

**Reply by rstrouse** [reply]
Your commit dates all the way back to May 30th.  You need to pull njsPC and dashPanel then rebuild your image.

**Reply by MaxVonEvil** [reply]
Right, so I just did that:

//before:
```
docker images
REPOSITORY                             TAG       IMAGE ID       CREATED       SIZE
msmi/nodejs-poolcontroller             latest    e66cdfe5d38c   6 weeks ago   196MB
msmi/nodejs-poolcontroller-dashpanel   latest    608ce4b993f7   7 weeks ago   168MB
```

//after
```
docker images
REPOSITORY                             TAG       IMAGE ID       CREATED       SIZE
msmi/nodejs-poolcontroller-dashpanel   latest    201627013745   5 hours ago   168MB
msmi/nodejs-poolcontroller             latest    1e1018d2af8b   5 hours ago   198MB
msmi/nodejs-poolcontroller             <none>    e66cdfe5d38c   6 weeks ago   196MB
msmi/nodejs-poolcontroller-dashpanel   <none>    608ce4b993f7   7 weeks ago   168MB
```

//cleanup old images
```
docker image rm e66cdfe5d38c
docker image rm e66cdfe5d38c
```

Starting up the containers again and loading the dashpanel, gauging by the commit numbers, it looks like both containers have updated:

![image](https://user-images.githubusercontent.com/19863636/179624774-3d37bc88-a433-458f-8a6b-6dfc73a6607b.png)

As you can see, the burgermenu and gear-menu's work fine, yet when I click anywhere in the marked area at the top, still nothing happens? Can you think of a good way to troubleshoot this?

![image](https://user-images.githubusercontent.com/19863636/179625563-ebd5757f-ef09-489f-a9dd-2dd433c8cb9d.png)

**Reply by rstrouse** [reply]
Hard refresh your browser cache.

**Reply by rstrouse** [reply]
Clicking on the Ready text should bring up the following dialog.
![image](https://user-images.githubusercontent.com/47839015/179630838-f7427fd5-20d8-4160-a57f-65d3a5b6337d.png)

**Reply by MaxVonEvil** [reply]
@rstrouse Augh! So much work and it was the last inch that tripped me up. Hard refresh was EXACTLY it! 

Thanks - Just one more question then. What would the API url+payload look like to change service mode? Does not look like it was in the #508 thread nor in Swagger?

**Comment by rstrouse** [comment]
The timeout value is required for timeout mode.

`PUT state/panelMode { "mode": "auto | service | timeout", "timeout": seconds }`

**Reply by MaxVonEvil** [reply]
Nice, I'll give it a try tonight - Thanks for the help!

**Reply by tagyoureit** [reply]
@MaxVonEvil you finally get a single command to turn off all features. https://github.com/tagyoureit/nodejs-poolController/issues/262

If you set a 1s service timeout that would effectively be this single button to rule them all.

**Reply by MaxVonEvil** [reply]
Yep and thanks! I am busy rewriting the old NodeRed control code, it's getting a lot cleaner, tossing out all the old repeated messages. (my button - my pressscious!!! lol)

**Reply by MaxVonEvil** [reply]
So, I ran a test and it works - except not quite the way I was hoping:
When I engage service or timeout mode, njspc indeed does relinquish control of the pump, however it still stops the pump before doing so, which is the behavior I am trying to avoid. 

Desired state is the pump reverts to it's onboard programmed schedule, which then can be interrupted by the pooltech. This still happens after ~30sec if I physically yank the plug out, i.e. sever the data connection.

As mentioned I had that (perhaps unintentional) behavior using PM with a native node install in the beginning, yet if within the realm of the possible, I would really love an option in njspc to intentionally mimic this behavior again - without having to resort to a hardware hack :)

**Reply by tagyoureit** [reply]
Can we do a quick test?

Can you change this line of code (on your local install) https://github.com/tagyoureit/nodejs-poolController/blob/51155cd8548bc278a242bc8568e939d07b67626f/controller/nixie/pumps/Pump.ts#L531

to be `try { if (!this.closing && pt.name !== 'vsf') await this.setPumpFeature(6); } catch (err) { };`

We took this out a while ago because we determined we don't need to set the pump program (and it was actually interfering with the operation of some VS pumps).  But maybe this will keep the program running in your case.  Make sure you either use `npm start` or manually compile the typescript `npm run build` if you are launching the app another way.

I'm also on the fence about including something like this in the app.  I want to know, for sure, that when I turn on service/timeout that my pump(s) will not keep running.  You are essentially asking us to be able to bypass a critical security feature and I don't feel comfortable in doing that.  However, if you can modify the code in this manner and it works than you have control over this on your own.

**Reply by MaxVonEvil** [reply]
Great, thanks for your willingness to try this out. I do see where you're coming from in terms of security. I guess my setup is just a smidge different. 

Anyhow I made the change inside the docker image, however it didn't seem to have any effect: When I entered service-mode from the dashpanel, the pump still stops.

That being said, this is where my docker-savvyness comes to a screeching halt, as I'm not sure if one needs to do some kind of rebuild after the edit. Here's the steps I took 

```
docker exec -u 0 -it <docker-image-id> /bin/bash
vi /app/controller/nixie/pumps/Pump.ts
#chamge the line as per your recommendation, save and exit
docker stop poolcontroller
docker start poolcontroller

```
However, when I went outside to check last night, the pump is still being stopped when servicemode kicks in. If you have any other suggestions or things you would like to test, let me know.

**Reply by tagyoureit** [reply]
Based on what I'm seeing from the [docker config files](https://github.com/tagyoureit/nodejs-poolController/wiki/Docker#build-a-customized-docker-image) (and I'm not an expert), the image is pre-compiled.

You can make the change directly in the compiled JS file.  `/dist/controller/nixielpumps/Pump.js`. The same line will look like:
```
                try {
                    if (!this.closing && pt.name !== 'vsf' && pt.name !== 'vs')
                        await this.setPumpFeature(6);
                }
                catch (err) { }
```
If you make that same change here you won't need to worry about compiling and a restart of the container should pick up the changes.

**Reply by rstrouse** [reply]
When you engage the timeout it releases the pump so that the display should not say DISPLAY NOT ACTIVE anymore.   This means that you can now set the speed at the pump.  When the timeout expires njsPC will again take control of the pump.  The same thing happens even if the messages stop after several seconds.  At that point the pump programming should be in control and pressing any one of the speed 1-4 should make it start.

**Reply by MaxVonEvil** [reply]
Gents, thanks for the suggestions
@tagyoureit, I made the edits in `/app/dist/controller/nixie/pumps/Pump.js` at line 561

![image](https://user-images.githubusercontent.com/19863636/180155399-22ff55e3-5600-475f-9ea0-9f75e247920c.png)

I restarted the container, unfortunately with no effect. I did notice in the same folder there is a Pump.js.**map** file, which I suspect is the minimized/"compiled" version, so I went into the container and renamed this file, and restarted the container once again.

This produced an effect, just not the one I was looking for. Let me break down the sequence of observed events:

1) pump is under control of njspc, running at 2000 rpm
2) servicemode is entered and pump stops immediately (still)
3) now, after the above changes there is a period about 30 seconds with 'DISPLAY NOT ACTIVE' on the pump as njspc is in service mode, i.e. not responding on any hailing frequencies.
4) after that the pump shifts over local mode, however it is still in stopped mode, which as discussed I am trying to avoid. I need it to resume local scheduling.

I can still obtain the required behavior by pulling the plug on the pump as it then will resume what ever locally programmed schedule is active at that time, however it seems njspc is still hellbent on stopping the pump. Only difference is now that the code change above seems to pretend the connection being broke (i.e. no immediate handover to the pump) yet with the preceding stop-command stil being sent, this doesn't make much difference unfortunatly.

I noodled over your concern about security and equipment and as I said, I understand the security/safety concern as a precusor to the reluctance implement an option for this. However as I think about it, all it takes is a data plug falling out and a scheduled speed still to be present on the pump to pave the way for unforeseen behavior, which you would not be able to do anything about from the control side, since the connection is severed. In the scenarios you reference where safety is a concern, one could argue that the real security would be based on all locally programmed speeds being set to manual or disabled.

@rstrouse - thanks for explaining the timeout option. I did test it and it, however it does unfortunately not resolve the issue with the pump stopping. I need the pool-tech to find the system as if it's being run locally, i.e. when he shows up, the pump is not being remote controlled and is running happily, not stopped.

So, are there any other code-snippets you would have me try?
Thanks again for a great project.
//Max

**Reply by tagyoureit** [reply]
The .map files only help debuggers step through the TS code when it the app is running the compiled JS code.

I'm traveling this week.  Instead of having you jump through some hoops which may or may not help you, I'll give this some more thought and try to test a couple things next week.

**Reply by rock-crusher** [reply]
Hi MaxVonEvil,

Your post here caught my eye and I couldn’t help but wonder why it is so important that the pool-tech believe/think that your pool is being controlled locally and is not automated. 

Is he one of those pool guys that expect to have sole control of your pool?

Just curious as to why the new service/timeout mode feature wouldn’t suffice in a situation like yours having a weekly pool service.

**Reply by rstrouse** [reply]
You know we are simply releasing control.  I wonder if you set njsPC to service mode and hit 1 on the pump if it starts.  Then if you set njsPC to auto and give it some time to take over again (DISPLAY NOT ACTIVE).   Then hit service mode again if it would go back to where it was on the pump.   I do know that if we stop sending messages regardless of whether the cable is connected the pump release control and do whatever it does without any automation.

I wonder if incidentally you turned off the manual program on the pump.

**Comment by MaxVonEvil** [comment]
@rock-crusher, it's part of the lease agreement. The legalese roughly translates to "Derp shalt not f***eth with poolgear", so here's le me doing just that, lol.

@rstrouse, yes that is correct; once njspc releases control, the pump will start if you hit start/stop, or any speed button. My problem is that no matter what I do (except severing the data connection) will leave the pump in a 0 RPM stopped state, where the desired state is to have it resume local programming. 

Last I am not sure what you mean by turned off the manual program on the pump? What I can tell you is that I verified the local programming and there are 3 active schedules which run back-to back, i.e. the pump never has an off period.

@tagyoureit, thank you! - as always your efforts are very appreciated. 

As I began the topic with, worst comes to the worst and we cant find an elegant solution, I can always resort to a pi-controlled relay, cutting one of the data wires, but I hope we don't have to go that route as it's what is generally referred to as a fugly hack ;)

**Reply by tagyoureit** [reply]
It's definitely possible to select a speed/program on the pump without going into remote control mode.  I can either find a way to do it with the current RS485 commands or possibly give you a script that you could run separately (or maybe an [advanced trigger](https://github.com/tagyoureit/nodejs-poolController/discussions/527])).  Worst case scenario there are the [legacy pump commands](https://github.com/tagyoureit/nodejs-poolController/wiki/pumps) that could be utilized.  I have to check whether setting it to manual and/or selecting a program are needed to get what you want.

**Reply by MaxVonEvil** [reply]
Thanks! Looking fwd to what you find out. Safe travels btw.

**Reply by rstrouse** [reply]
Try this.  

1. Set njsPC to service mode.  
2. Press the 1 button on the pump (the pump should start running)
3. Set njsPC to auto and let it do what it does for half a minute or so.
4. Set njsPC to service mode and observe whether the 1 button is on.

**Reply by MaxVonEvil** [reply]
@rstrouse, pardon the deferred reply. Been out of town so I've been unable to test
At at step two above, behavior is different from what one might expect:

Pressing speed 1 will indeed show the associated speed programmed, however as the last thing njspc does is stop the pump, it does not start when the speed is selected. I can of course hit the start/stop button and it will start at that speed, but I guess that is not really what we are after here.

**Reply by rstrouse** [reply]
I am trying to get my head around what is different.  njsPC always issued a release.  What I am curious about is that if njsPC simply stops sending drive state messages to my pumps they shut down.  The reason I say that is because nothing makes sense here if the pump is operating as it is advertised.   A disconnect should return the control to whatever manual pump setting is currently selected.  It simply issues a command to release it.


So just to be sure you placed njsPC in service mode then pressed 1 on the pump.  After the pump started you then select auto in njsPC which now takes control of the pump.  When you place the njsPC back into service mode is the 1 still illuminated on the pump?

**Reply by MaxVonEvil** [reply]
Hi, without knowing the specifics of the command, the release is apparently being interpreted as a stop by the pump. 

To answer your question, entering into service mode from any state, the pump goes to displaying 0 RPM and it's stopped (the little orange led at the start/stop button is off)

Hope this helps, let me know if there are other items/tests you would like to see

**Reply by rstrouse** [reply]
I want you to change that and start the pump using the manual control panel then place it back into auto.  After njsPC takes back control place it in timeout or service and check to see if the button gets re-illuminated.

**Reply by MaxVonEvil** [reply]
Just to clarify, what specifically would you like me to change. 

Second, when you say illuminated it sound like you are referring to the buttons being backlit. This is not the case here, the only illumination is the pumps LCD display. FYI There is no display indication there that speed 1 was selected. It just reads 0 RPM STOPPED. My only course of action is to put it back in Auto mode or hit the start/stop button. ø manually

Timeout produces the same result, Albeit with it's timed delay

**Reply by rstrouse** [reply]
So you do not have a 1 2 3 4 button with an LED at the top?  What model do you have?

**Reply by MaxVonEvil** [reply]
While I do have the 1..4 buttons, no there is no LED illumination on or next to each button.

My pump is a Pentair Intelliflo VS model# 011018

**Reply by rstrouse** [reply]
Does your control panel on the pump look like this?
![image](https://user-images.githubusercontent.com/47839015/181049719-1ece87c0-7dca-42d6-870c-331330bfa559.png)

**Reply by MaxVonEvil** [reply]
It does, yet (after brushing off the dust) I did notice I have those little LED cutouts above the speed buttons, however those never light up on my pump. I'll send a picture in a bit

**Reply by rock-crusher** [reply]
Hi guys, I have been somewhat following this and decided to try what rstrouse has been asking of MaxVonEvil. I too have been playing around with the new service/timeout modes and thought to try this just to see how mine reacts in this same scenario.

> So just to be sure you placed njsPC in service mode then pressed 1 on the pump. After the pump started you then select auto in njsPC which now takes control of the pump. When you place the njsPC back into service mode is the 1 still illuminated on the pump?

rstrouse, I am getting the same results as MaxVonEvil. When doing exactly what you stated above and put back into service mode the display reads 0 RPM, it's stopped and the only LED that is on is the green power indicator.

BUT… This is how it is supposed to act correct?

**Reply by rstrouse** [reply]
Yes this is how it is supposed to act.  When you release control it looks like it sets the RPM to 0 which we definitely want to happen.  This tracks with what a typical Pentair Panel does.  The pump should be released and idle awaiting command from the local panel.

I think that Max would like the software to simply stop communicating which in his case does not stop his pump.  However, on my pumps if that happens they stop within a minute.  It may depend on the IntelliFlo version though because I have the original 4x160 pumps.

I wonder though if setting a program on the pump to be scheduled might start that schedule on the pump.  My pumps do not have those options.

**Reply by MaxVonEvil** [reply]
Gents, glad to hear that I'm not the only one seeing this.

As for how it is intended to act, I understand where Tag was coming from with security concerns. However if different use-cases such as ours are in scope, perhaps a configuration file entry to halt comms would be in the realm of the possible? This would to address those Pentair pump-firmwares which apparently behave subtle different ways.

Thanks
//Max

**Reply by rstrouse** [reply]
Hey since I cannot make what you are doing happen, can I get you to go into the RS485 Ports Tab and simply disable the port to see if it gets us there for your situation.

**Reply by tagyoureit** [reply]
FYI - I'm going to be traveling again this coming week and won't get a chance to look at this.  Still on my radar, though.

**Comment by MaxVonEvil** [comment]
I will try and do that. Unfortunately it may be a few days as we had a flood in the neighborhood and my pool looks like it's filled with chocolate milk, i.e. all my gear is offline until we can get the silt cleaned out of the pool. Will give a shout as soon as I get to try this

**Reply by rstrouse** [reply]
Yikes.  Hope you get it cleaned up soon and the flood waters recede.  Let me know if it works.

**Reply by rock-crusher** [reply]
Sorry to hear this Max, hope all goes well with the cleanup.

**Reply by MaxVonEvil** [reply]
Thanks guys appreciate it, this is what the neighborhood looked like last night and what the pool looked like this morning 
![IMG_20220727_221151](https://user-images.githubusercontent.com/19863636/181664701-5d44603d-cea8-461d-9b45-07e6b0862502.jpg)
![IMG_20220728_092011](https://user-images.githubusercontent.com/19863636/181664714-fb9a956a-d746-4d65-acb0-bd67e781faf5.jpg)

Instant frog pond, yuck.

**Reply by rstrouse** [reply]
Yeah that looks like a mess.  I see the sun in the pics so that is a good sign.

**Reply by rock-crusher** [reply]
Wow, that is a mess, I’m sure it will be drained, cleaned etc.

Excuse me Max, this box thingy, what is it and what’s its purpose, very interesting to say the least.

<img width="300" alt="Thingy" src="https://user-images.githubusercontent.com/94593954/181675402-e441deeb-c208-4237-98dc-1100fb53bc4b.png">

**Reply by MaxVonEvil** [reply]
Sure, It's a waterfall box painted to match the pipework. Wifey was tired of the regular double pipe splashing water out, as it was noisy as hell. The box delivers a nice flat flow which only makes a third of the splash sound (color me nerd, but I have a dB meter to back me up, lol). 

Anyway, you can find the box here https://www.lowes.com/pd/smartpond-Black-Pond-Waterfall-Box/1000598779

I've got another project on the back burner to construct a set of laminar flow jets, but that's a story (and topic methinks ;) for another day.

**Comment by MaxVonEvil** [comment]
@rstrouse, so hey - finally got the backyard and pool rehabilitated as of yesterday (besides flushing out the pool and all the pipes a shopvac and a pressurewasher set to stun did wonders) 

Anyway, I got around to trying your suggestion of manually disabling the RS485 interface in njspc dashboar. It actually works! After about 45 sec the pump falls back to it's local programming, i.e. it does not stop, yay!

So, I guess I need to understand how to do that either through a commandline or the REST API. I checked the [swagger](https://tagyoureit.github.io/nodejs-poolcontroller-api/) site, may be somewhat outdated, yet there is no obvious mention of any method to change the RS485  port, yet since one can do it from the dashboard, there's obviously something that can do it?

Anyway, if you guys can give me a kick in the right direction it would be appreciated as always!
Thanks,
/Max

**Reply by rstrouse** [reply]
When I get back in town I’ll look it up.  In the mean time you can see exactly what dashPanel is sending from the browser console in chrome.

**Reply by tagyoureit** [reply]
This is what @rstrouse is talking about.

![2022-08-06_19-48-32](https://user-images.githubusercontent.com/1646840/183272881-c0d5447f-e9c2-4afa-921a-a261f19f69e1.png)

If you only have one port it will be portId 0 and you can send a PUT to `/app/rs485Port` with a payload of `{'portId': 0, 'enabled': true|false}` (you don't need to add any of the other settings).  

I didn't get to look at the older commands for setting the pump to the right state, and as long as this works for you I won't bother.  Glad we found a solution.

**Reply by MaxVonEvil** [reply]
@tagyoureit, @rstrouse 
Gents, thank you both for your willingness to work with me through this conundrum. I made a couple of NodeRed items the URL's you provided above, so others might benefit from this. They look like this and the NodeRed JSON import can be found [here](https://pastebin.com/tusavNHf).

![image](https://user-images.githubusercontent.com/19863636/183311334-2ca051ff-8647-4b5a-8512-a5b2a51ea8d0.png)

**Comment by MaxVonEvil** [comment]
One minor item, this might be either rather generic or the wrong place to ask, but here goes:

As per your suggestion, I was looking in the browser's debug console and sure enough, I can see all the messages from the pump, the ones I am interested in is the wattage and and rpm's: 
![image](https://user-images.githubusercontent.com/19863636/183312242-be23e08f-ad92-42f1-89a2-00d5f41bbc94.png)

As things are now, I subscribe to these values via MQTT, yet since I'm doing everything else through the API, it makes sense to somehow subscribe to the the messages above and eliminate MQTT altogether for this project. Problem is I don't know enough about API subscriptions if there is such a thing. 
Figuring it out, I can probably find the equivalent method of doing subscribing in NodeRed.

Am I limited to doing a interval polling like GET /state/pump/50 and then pulling the watts and RPM's, or is there a smarter way?

**Reply by rstrouse** [reply]
You subscribe to the web socket.  It will emit to your client whenever there is a change.  Simply make a socket io connection to njsPC and trap the events that you want.  Everything in njsPC emits on a socket.

**Reply by MaxVonEvil** [reply]
Thanks, so I did a little digging how to do it. 
Node Red comes with a nice websocket client, which can be configured with the websocket url to listen to:

![image](https://user-images.githubusercontent.com/19863636/183370464-27b32b62-c8c1-4faf-88b9-766ad5104fad.png)

I've tried several combinations around ws://my.njspc.ip.here:5150/websocket.io however the client does not seems to latch on and remains disconnected. Any suggestions as for what the URL needs to look like?

Thanks,
Max

**Reply by tagyoureit** [reply]
5150 is the port for dashPanel. NjsPC communicates on port 4200.

**Reply by MaxVonEvil** [reply]
Right, I should have mentioned I tried that port as well. 
Is there any particular path which needs to be added after the port number, like /websocket.io  /app etc? I've tried several combinations without success. The node just sits disconnected:

![image](https://user-images.githubusercontent.com/19863636/183460641-360ac589-fe21-477d-bdb2-1a15337e3d74.png)

Btw, to get a second opinion, I found a websocket test app and it's also not giving me any love. I suspect I am missing some specific path:

![image](https://user-images.githubusercontent.com/19863636/183474346-98a25137-7f31-4cce-83ae-f5dfc2051f49.png)

I've tried /websockets.io /sockets.io /app /status /status/all and just the host. Kinda running out of ideas :)

**Reply by rstrouse** [reply]
The address should simply be `ws://10.2.2.3:4200` if that is where your njsPC is located.  Then this will return socket data for all items on njsPC with an event identifier for instance of `pump`, `circuit`, `feature`, ... etc.

**Reply by MaxVonEvil** [reply]
Strange... That was one of the first ones I tried. The websocket test utility I'm using just returns this:

```
8-8-2022 10.33.57  Try to connect ws://10.2.2.3:4200
8-8-2022 10.33.58  [E] Connect exception: The text associated with this error code could not be found.

The server returned an invalid or unrecognized response

			> ErrorHttpInvalidServerResponse
			> Unable to connect to the remote server
```

Per chance would there anything that needs to be tweaked in the njspc config.json? I took a peek just now yet I do not recall changing anything. This is the only part I found referencing sockets:

```
252         "socket": {
253           "transports": [
254             "websocket"
255           ],
256           "upgrade": false,
257           "reconnectionDelay": 2000,
258           "reconnection": true,
259           "reconnectionDelayMax": 20000
260         },

```

**Reply by tagyoureit** [reply]
If your client is web based, you could be running into CORS issues.  I tried with a web based app and ran into that.  I tried with Postman (local) and it connected.  

![2022-08-08_10-44-29](https://user-images.githubusercontent.com/1646840/183480563-37d6d12e-46b6-4649-a661-8ef3bdf4e011.png)

You may want to explore some of the Socket.IO specific extensions if the generic websocket doesn't work.

**Reply by rstrouse** [reply]
Nope the websocket client is in control of what it needs as far as telling the server that it is connecting.  You mentioned earlier that you had a sockeio or websocket client in Node Red.  You will likely want the socketio client since it is running node.

**Reply by MaxVonEvil** [reply]
@tagyoureit, so the test app I mentioned I had found was this one: https://www.microsoft.com/store/productId/9P65MNB62734 - However, I gave PostMan a whirl as well. Connecting to ws://10.2.2.3:4200 this is what it came back with the following (tldr; that dog ain't barking either)
```

Could not connect to ws://10.2.2.3:4200
10:56:16
Error: socket hang up
Handshake Details
Request URL: http://10.2.2.3:4200/
Request Method: GET
Request Headers
Sec-WebSocket-Version: 13
Sec-WebSocket-Key: blhablahblahblah==
Connection: Upgrade
Upgrade: websocket
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
Host: 10.2.2.3:4200

```
@rstrouse So I gather so there's a difference between websockets, and Sockets.IO. I found a Node Red Sockets.IO extension [here](https://flows.nodered.org/node/node-red-contrib-socketio-client) which I'm going to take for a spin and see what it does. That said, I am somewhat puzzled that PostMan can't talk to the controller?

**Reply by rstrouse** [reply]
Are you sure that 10.2.2.3 is the IP address for njsPC?

Yes the data payload for socket.io and websocket.io are different.  The latter is much less robust and needs to be manually parsed.

**Reply by MaxVonEvil** [reply]
Yes, pretty sure as dashpanel and controller live on the same Pi:
![image](https://user-images.githubusercontent.com/19863636/183507742-57f9dee2-e7bb-47a8-a938-81afcdcc02a1.png)

Other than that, I've been unsuccessful in finding a sockets.io node for Node Red as many repos have not been updated in months or years. Many of these are coded to support only up to Node 14 and now with Node Red 13.x we are on Node 16. All that said, I would still have expected postman to work?

Could it have something to do with that I run NjsPC in a docker container?

**Reply by rstrouse** [reply]
I don't have any experience with docker.  But I have to say that for something that is supposed to make things run reliably all the time it seems to be one of the largest thorns in this developers side.  I suppose applications that are not built cross-platform may benefit but between the permission issues and access to other resources it appears to be a PITA.  Either way I do not believe that docker is the issue in this case.

For Postman update your postman client.   There should be a dropdown on the upper left that indicates whether you are using a raw or Socket.IO socket.  You want Socket.IO.  The beta version of the Postman socket implementation didn't retain the Socket.IO agent.

**Reply by MaxVonEvil** [reply]
Hi guys, been away for some days as hacker-week is winding down here in Vegas.
Anyway, just to follow up on a few loose ends:

**Docker:** 
Couldn't agree more, wish it was easier - however having multiple Node versions on the Pi, global and local conflicting isn't fun either.

**Postman:** 
I believe I am running the latest postman; just pulled down v9.28.2 to be sure. As you can see in the screenshot below it is connecting, however it seems I have to create a listener for it to receive messages, yet it seems no matter what I type as a name, I do not receive anything?

![image](https://user-images.githubusercontent.com/19863636/184556199-61fc6ced-5b45-423e-ae8b-9aa584db582a.png)

**Proxy oddity:**
Anyhow, I remembered an oddity from when I originally installed njspc. When I set up the connection between the dashpanel and the server, I had to check the Use Proxy to NjsPC Server in order for it to connect.

![image](https://user-images.githubusercontent.com/19863636/184553679-d52b4c5b-d3c1-405a-b231-9bb26a3cd98a.png)

If I uncheck the proxy box, the dashpanel finds the server, yet it won't connect:

![image](https://user-images.githubusercontent.com/19863636/184554598-37229ba1-22ff-4979-bee3-b944c5122af4.png)

This seems somewhat weird as both components are running on the same pi, albeit in different docker containers - is that the purpose of the proxy function per chance? Just wondering out loud if it's docker giving more grief?

**Nodered:**
So with the understanding that it's a sockets.io listener I am after, I found one [here](https://flows.nodered.org/node/node-red-contrib-socketio-room). Unfortunately the developer of said node has not updated it in months and it's relying on an older version of node. I am currently waiting for an update there.

**Reply by rstrouse** [reply]
Try using the IP address 172.22.0.3.  Very odd that the ip is so different than it was previously.  A completely different local net.

**Comment by MaxVonEvil** [comment]
Thanks @rstrouse - I looked further into this and it seems to be all Docker... (sorry Tag). 

It looks like it creates a unique IP for each container, this is where the 172.22.x.x addresses are coming from. while the address can be pinged when SSH'ed locally into the Pi where the containers are running, one can't reach these addresses from outside, unless we start modding routing tables, which should not be necessary. It seems docker automatically adds the internal routing table entries. Besides when you hit the Pi's wlan0 ip address on port 4200 it gets routed to the docker container portmapping configured in the docker-compose.yml

If you examine the screenshot above (big white one) you can see that Postman (running on another computer) actually connects to njsPC: 

![image](https://user-images.githubusercontent.com/19863636/184573467-d0250779-68cb-4888-b509-1b20e366c703.png)

It might very well be that I don't know Postman well enough, yet it seems I need to add a listener. 
Does this postman listener need to have a specific name to pick up messages from njspc?

**Reply by tagyoureit** [reply]
If you go back to the javascript console you can see the events in real time as they are emitted.
`circuit`, `body`, `temps`, `controller`, etc are all valid events.

![image](https://user-images.githubusercontent.com/1646840/184581093-1d744312-82cd-47d7-b4f6-9b555b437932.png)

You may want to check out [this article](https://www.freecodecamp.org/news/how-to-get-a-docker-container-ip-address-explained-with-examples/) among others that discusses container networking principles.  Further, the 172.[16-31] address block (along with 10 and 192.168) is private.  You will never be able to route that to another network.  Typically, you can map ports to the container and that will expose the internal ip on the external ip:port combination.  See the [example docker config files in the wiki](https://github.com/tagyoureit/nodejs-poolController/wiki/Docker).

**Reply by MaxVonEvil** [reply]
Hi, I did go back and had a look and this is the part I'm having a heckuva time figuring out:
Yes - I get the messages rolling in nicely through FireFox's debug console, just like you do in Chrome

Summoning the vast powers of deduction, I'm guessing this means it's not a docker issue - Elvis IS leaving the building, so to speak :)

My end-state goal here is to be able to pick up the messages via a socket.io listener in Node-Red, which is obviously beyond the scope of our conversation here, as I am working with another node developer on that.

However what really puzzles me is that beyond the debug console in the browser, I have been unable to find any sockets.io listener - PostMan included - which can pick up the sockets.io messages from njspc? 

Anyway, worst comes to the worst I still have MQTT to fall back on for sensor readings, albeit it tickles my OCD that I can't get rid of it as I've got everything else running by way of the API.

**Reply by rstrouse** [reply]
You must upgrade your postman app.  It will be in the upper right hand corner of the display as a dropdown.
![image](https://user-images.githubusercontent.com/47839015/185232144-1704614e-a48e-4769-b3b6-de7142424c50.png)

**Reply by rstrouse** [reply]
Actually I meant 3 rights which would have been left.

**Reply by MaxVonEvil** [reply]
@rstrouse, pretty sure I'm on the latest and greatest Postman - If you look at my screenshot I posted earlier [here](https://github.com/tagyoureit/nodejs-poolController/discussions/546#discussioncomment-3394548), you'll notice that I'm in fact connected to Sockets.io (upper right) to ws://ip:4200 (lower left) and a handshake has happened. 

However **I do not have any listensers configured** in the lower right, as I am not sure how to configure those?

I can see from the [screenshot earlier](https://github.com/tagyoureit/nodejs-poolController/discussions/546#discussioncomment-3351596) from @tagyoureit that he's got 3 listeners configured, but not what they look like. Any chance you can share how those are configured? Thanks!

**Reply by rstrouse** [reply]
Oh you need a listener.  The listener will be any of the emits you see in the browser console.  Simply click on the listeners at the lower right (got the correct hand this time) that says Listeners.  Then enter any event name you want to listen to.

Some valid selections include:
controller
temps
body
circuit
feature
circuitGroup
lightGroup
schedule
pump
heater
valve
chemController
chemDoser
chemDose
chlorinator

![image](https://user-images.githubusercontent.com/47839015/185274608-2c1c90dc-0df8-4e54-ab8b-eec18d505c2b.png)

**Reply by MaxVonEvil** [reply]
Ah-HA! it was exactly those magic words I was looking for:
![image](https://user-images.githubusercontent.com/19863636/185278477-3e79b95c-d040-4429-bfd8-e2fd57c39075.png)

While I knew where to put them, I think I might have gotten close as I tried with temp, but it was in PLURAL, curses! 😂

Anyway I just need to find a working socket.io node for NodeRed and I'm golden, but that's a conversation for another forum. Thanks for the help!

**Reply by johnny2678** [reply]
> Anyway I just need to find a working socket.io node for NodeRed and I'm golden

Please post back here if you do!
